clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +348,560 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+        log.info("Revoking previously assigned partitions {}", Utils.join(revokedPartitions, ", "));
+
+        logPausedPartitionsBeingRevoked(revokedPartitions);
+
+        // Mark partitions as pending revocation to stop fetching from the partitions (no new
+        // fetches sent out, and no in-flight fetches responses processed).
+        markPendingRevocationToPauseFetching(revokedPartitions);
+
+        // Future that will complete when the revocation completes (including offset commit
+        // request and user callback execution)
+        CompletableFuture<Void> revocationResult = new CompletableFuture<>();
+
+        // Commit offsets if auto-commit enabled.
+        CompletableFuture<Void> commitResult;
+        if (commitRequestManager.autoCommitEnabled()) {
+            commitResult = commitRequestManager.maybeAutoCommitAllConsumed();
+        } else {
+            commitResult = CompletableFuture.completedFuture(null);
+        }
+
+        commitResult.whenComplete((result, error) -> {
+
+            if (error != null) {
+                // Commit request failed (commit request manager internally retries on
+                // retriable errors, so at this point we assume this is non-retriable, but
+                // proceed with the revocation anyway).
+                log.debug("Commit request before revocation failed with non-retriable error. Will" +
+                        " proceed with the revocation anyway.", error);
+            }
+
+            CompletableFuture<Void> userCallbackResult = invokeOnPartitionsRevokedCallback(revokedPartitions);
+            userCallbackResult.whenComplete((callbackResult, callbackError) -> {
+                if (callbackError != null) {
+                    log.error("User provided callback failed on invocation of onPartitionsRevoked" +
+                            " for partitions {}", revokedPartitions, callbackError);
+                    revocationResult.completeExceptionally(callbackError);
+                } else {
+                    revocationResult.complete(null);
+                }
+
+            });
+        });
+        return revocationResult;
+    }
+
+
+    /**
+     * Make new assignment effective and trigger onPartitionsAssigned callback for the partitions
+     * added.
+     *
+     * @param assignedPartitions New assignment that will be updated in the member subscription
+     *                           state.
+     * @param addedPartitions    Partitions contained in the new assignment that were not owned by
+     *                           the member before. These will be provided to the
+     *                           onPartitionsAssigned callback.
+     * @return Future that will complete when the callback execution completes.
+     */
+    private CompletableFuture<Void> assignPartitions(
+            SortedSet<TopicPartition> assignedPartitions,
+            SortedSet<TopicPartition> addedPartitions) {
```

Comment:
The legacy coordinator does trigger the `onPartitionsAssigned` with empty partitions (not the `onPartitionsRevoked` though), so I intentionally left the same behaviour, makes sense? I had also added a note on the [invokeOnPartitionsAssignedCallback](https://github.com/apache/kafka/blob/4a416a2ea7bc9a0d4f1ca1f8736cada784127b1a/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L859) to make sure that we keep that contract when implementing callbacks.

Reply:
Interesting... I wonder if this was done on purpose or if this just a bug. I don't really see the value in calling `onPartitionsAssigned` without any partitions. I suppose that we should double check this. We could perhaps file a JIRA and clarify this separately. What do you think?

Reply:
Agree on handling that separately, but leaving the current behaviour as in the legacy coordinator. I also don't see the value either but it would introduce a change on when the `onPartitionsAssigned` is called or not, so let's put some more though on it. Filed [KAFKA-15843](https://issues.apache.org/jira/browse/KAFKA-15843) to follow-up on this.